### PR TITLE
Raw results and metadata

### PR DESCRIPTION
**Pull request by thisac**
**Context:**
There's no way to access non-samples results (i.e., metadata) in the `Result` class.

**Description of the Change:**
* A `Result.raw` method is added that returns the raw result dictionary.
* A `Result.metadata` method is added that returns the metadata of the results (i.e., everything except the `"output"` entry).
* Tests are added for the `Result` class.
* A setter method is added for `Result.state` allowing a state to be set iff a state has not previously been set and the `Result` object contains `samples_dict` samples, indicating that it's a local simulation.

**Benefits:**
* Results metadata can be accessed without using the private `Result._result` method.
* The `Result` class is properly tested.
* The state can be properly set in the result object.

**Possible Drawbacks:**
* The state setter relies on the fact that all local simulations have a `samples_dict` (which is currently the case).

**Related GitHub Issues:**
None

### CODE REVIEW

File: strawberryfields/result.py [hunk 1]
```diff
@@ -73,6 +80,15 @@ def __init__(self, result: Mapping, ancilla_samples: Mapping = None) -> None:
...
         self._ancilla_samples = ancilla_samples
 
+    @property
+    def raw(self) -> Mapping:
+        """The raw results dictionary containing any samples and metadata.
+
+        Returns:
+            dict: samples and metadata
+        """
+        return self._result
```
Comment (thisac) [inline strawberryfields/result.py:90]: I was thinking about whether it would be better not to return the results directly like this, but rather copy it or simply return a view of the dictionary, to avoid users accidentally changing anything in the `Result` object. Not sure if it's worth it or, in that case, how to best do that.

File: tests/api/test_result.py [hunk 2]
```diff
@@ -16,30 +16,99 @@
...
 pytestmark = pytest.mark.api
 
 
+raw_results = {
+    "output": [np.ones((2, 3, 4, 5))],
+    0: [1, 2, 3],
+    1: [4, 5],
+    "other_data": [1, 0, 1, 2],
+    "string": "metadata",
+}
```
Comment (Mandrenkov) [inline tests/api/test_result.py:32]: I know this is just a test module, but there are a few issues with this result:
1. The keys `0` and `1` are illegal since all keys must be strings.
 2. The list values are illegal since all values (with the exception of the `"output"` value) must be NumPy arrays.
 3. The string value `"metadata"` is illegal (see previous point).

File: tests/api/test_result.py [hunk 3]
```diff
@@ -16,30 +16,99 @@
...
+
+    def test_raw(self):
+        """Test that raw results are correctly returned."""
+        result = Result(raw_results)
+        assert result.raw == raw_results
+
+    def test_samples_dict(self):
+        """Test that ``samples_dict`` is correctly returned."""
+        result = Result(raw_results)
+        assert result.samples_dict == {0: [1, 2, 3], 1: [4, 5]}
```
Comment (Mandrenkov) [inline tests/api/test_result.py:52]: I just realized that SF v0.20.0 may not be handling the samples dictionary correctly - the `Result` class should be checking for _strings_ in the raw result which resemble integers.

File: tests/api/test_result.py [hunk 4]
```diff
@@ -16,30 +16,99 @@
...
+
+    def test_ancilla_samples(self):
+        """Test that ancilla samples are correctly returned."""
+        ancilla_samples = {0: [0, 1], 2: [1, 3]}
+        result = Result(raw_results, ancilla_samples=ancilla_samples)
+        assert result.ancilla_samples == ancilla_samples
+
+    def test_state(self):
+        """Test that the state can be correctly set and returned."""
```
Comment (Mandrenkov) [inline tests/api/test_result.py:61]: This also tests that a state cannot be set twice (unless, of course, the state is set to `None`).

**Review by Mandrenkov — COMMENTED** [review]
Looks great! 🌟

My only concern is regarding the types of the raw result keys and values.

File: strawberryfields/result.py [hunk 5]
```diff
@@ -146,9 +175,25 @@ def state(self) -> Optional[BaseState]:
...
+
+        Raises:
+            TypeError: if state is already set
+            ValueError: if result have ``samples_dict`` (i.e, do not come from a local simulation)
+        """
+        if self._state:
+            raise TypeError("State already set and cannot be changed.")
+        # Samples from local simulation jobs will always have a samples_dict containing the same
+        # samples organized by measured mode. `self.samples` is either `None` or an array.
+        if self.samples is not None and len(self.samples) != 0 and not self.samples_dict:
```
[the remark this answers is not in the document]
Comment (thisac) [reply strawberryfields/result.py:190]: This one is a bit tricky. If `self.samples` is an ndarray, casting it to a bool raises an error,
```pycon
>>> bool(np.array([1, 2, 3]))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()
```
so using only `if self.samples ...` doesn't work, and using `len(self.samples)` doesn't work if it's `None`. I agree that this looks a bit ugly, but I couldn't figure out a better way to do this. Do you have any other ideas?

A workaround could be to use `if np.any(self.samples) ...` which technically works, but isn't supported by type hinting if `self.samples` is `None` (which causes annoying squiggles to pop up in my editor), or `if np.any(self.samples or []) ...`, but I'm not too fond of either of them.

File: tests/api/test_result.py (hunk 2 above)
Comment (thisac) [reply tests/api/test_result.py:32]: I'm not sure we ever decided that keys must be strings and values must be arrays, or did we? At least not on the SF side. The idea behind the `samples_dict` was to check for integer keys and have them correspond to modes, and I though metadata could potentially be strings (if needed).

File: tests/api/test_result.py (hunk 3 above)
Comment (thisac) [reply tests/api/test_result.py:52]: The usage of integers as keys was deliberate here. See my other comment above.

File: strawberryfields/result.py (hunk 1 above)
Comment (thisac) [reply strawberryfields/result.py:90]: Sure, I agree. 👍

File: strawberryfields/result.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (Mandrenkov) [reply strawberryfields/result.py:190]: Ah, okay! I see what you mean. In that case, I agree that the current implementation is the way to go.

File: tests/api/test_result.py (hunk 2 above)
Comment (Mandrenkov) [reply tests/api/test_result.py:32]: Actually, this restriction is enforced by the .npz format which stores data in the form `dict[str, np.ndarray]`.

File: tests/api/test_result.py (hunk 3 above)
Comment (Mandrenkov) [reply tests/api/test_result.py:52]: Let's continue this discussion in the above comment!

File: tests/api/test_result.py (hunk 2 above)
Comment (heltluke) [reply tests/api/test_result.py:32]: Yeah, real data will never look like this. For example, on QKD_01, raw results presently look like
```
{'alphas': array([-0.00023017-1.17238805e-03j,  0.00163755+1.69466410e-03j,
         0.00170546+1.58876506e-03j, ..., -0.00131527+1.67085418e-03j,
         0.0005287 -7.48599295e-05j,  0.00027381+2.05610324e-04j]),
 'output': [array([ 9.29057598e-04,  1.29526854e-03,  2.39390135e-03, ...,
         -1.32498145e-03,  1.77919865e-05, -1.01980567e-03])]}
```
where "alphas" is the metadata coming from the `.npz` (that we save with the key "alphas").

File: tests/api/test_result.py [hunk 6]
```diff
@@ -49,19 +118,19 @@ def test_tdm_print(self, capfd):
...
 
-    def test_unkown_shape_print(self, capfd):
+    def test_unknown_shape_print(self, capfd):
         """Test that printing a result object with samples with an unknown shape
         provides the correct output."""
         samples = np.ones((2, 3, 4, 5))
         result = Result({"output": [samples]})
         print(result)
-        out, err = capfd.readouterr()
+        out, _ = capfd.readouterr()
```
Comment (heltluke) [inline tests/api/test_result.py:133]: If you wanted to save yourself the throwaway `_`, I note that the `pytest` examples seem to recommend writing this as `captured = capfd.readouterr()` and making assertions against `captured.out`.

File: strawberryfields/result.py (hunk 1 above)
Comment (heltluke) [reply strawberryfields/result.py:90]: Is there a clear use case for accessing the results like this?

**Comment by heltluke** [comment]
Gotta love `f-strings` :heart:

File: strawberryfields/result.py [hunk 7]
```diff
@@ -146,9 +175,25 @@ def state(self) -> Optional[BaseState]:
         """
         return self._state
 
+    @state.setter
+    def state(self, state) -> None:
```
[the remark this answers is not in the document]
Comment (thisac) [reply strawberryfields/result.py]: Added a check that the set state is either a `BasisState` or `None`, simplifying the process of attempting to add a non-existing state which was being made in SF.

File: tests/api/test_result.py (hunk 6 above)
Comment (thisac) [reply tests/api/test_result.py:133]: Good idea! Updated!

File: strawberryfields/result.py (hunk 1 above)
Comment (thisac) [reply strawberryfields/result.py:97]: Maybe not. It'd just be a way of easily getting direct access to the results obtained from XCC. 🤔

Comment (Mandrenkov) [reply strawberryfields/result.py:97]: Perhaps it could be useful for debugging? But then again, users can just access `_result` directly.

File: strawberryfields/engine.py [hunk 8]
```diff
@@ -221,6 +221,9 @@ def _run_program(self, prog, **kwargs):
             commands that were applied to the backend, the samples returned from the backend, and
             the samples as a dictionary with the measured modes as keys
         """
+        # `_run_program` is called in `BaseEngine._run` but should never be called
+        # from a `BaseEngine` object, in which case an error should be raised.
+        raise NotImplementedError("'run_program' method must be implemented on child class")
```
Comment (Mandrenkov) [inline strawberryfields/engine.py:226]: Are we sure `abc.abstractmethod` won't raise something similar already?

File: strawberryfields/result.py (hunk 1 above)
Comment (heltluke) [reply strawberryfields/result.py:97]: This :point_up: was exactly my thinking

File: strawberryfields/result.py [hunk 9]
```diff
@@ -75,17 +75,24 @@ class Result:
...
         """The raw results dictionary containing any samples and metadata.
 
+        .. warning::
+
+            The raw values are returned by reference, meaning that any changes to
+            them will be reflected in the ``Result`` object.
+
         Returns:
-            dict: samples and metadata
+            Mapping[str, Union[np.ndarray, List[np.ndarray]]: samples and metadata
```
Comment (Mandrenkov) [inline strawberryfields/result.py:95]: Just curious: is the convention in PL and SF to use `list` or `List` in the Sphinx docstrings?
```suggestion
            Mapping[str, Union[np.ndarray, list[np.ndarray]]: samples and metadata
```

File: strawberryfields/engine.py [hunk 10]
```diff
@@ -374,19 +381,24 @@ def _run_program(self, prog, **kwargs):
...
+                    f"The operation {cmd.op} has not been implemented in {self.backend} for the "
+                    f"arguments {kwargs}."
                 ) from None
 
         # combine the samples sorted by mode, and cast correctly to array or tensor
         samples, samples_dict = self._combine_and_sort_samples(samples_dict)
+
+        if isinstance(prog, TDMProgram) and samples_dict:
+            samples_dict = reshape_samples(samples_dict, prog.measured_modes, prog.N, prog.timebins)
+            # transpose the samples so that they have shape `(shots, spatial modes, timebins)`
```
Comment (heltluke) [inline strawberryfields/engine.py:399]: Why not simply have the simulator return results in the desired form? There was certainly a bit of effort on the hardware side to get them in `(shots, sptatial modes, timebins)` format.

**Review by Mandrenkov — APPROVED** [review]
Some of the new additions are a little beyond my domain of expertise but the rest looks good!

One thing we should decide before merging is whether `Result.raw` is still needed.

File: strawberryfields/result.py [hunk 11]
```diff
@@ -80,7 +103,7 @@ def samples(self) -> Optional[np.ndarray]:
         Returned measurement samples will have shape ``(shots, modes)``.
 
         Returns:
-            array[array[float, int]]: measurement samples returned from
+            ndarray[ndarray[float, int]]: measurement samples returned from
```
Comment (heltluke) [inline strawberryfields/result.py:106]: Should this read `np.ndarray` or even `numpy.ndarray`?

File: strawberryfields/result.py [hunk 12]
```diff
@@ -95,22 +118,41 @@ def samples(self) -> Optional[np.ndarray]:
...
+            the prior measured sample on the same mode is replaced. The samples
+            dictionary, on the other hand, keeps _all_ measurements.
+
+        Returns:
+            Mapping[int, List]: mode index associated with the list of measurement outcomes
+        """
+        return self._samples_dict
+
+    @property
+    def metadata(self) -> Mapping[str, np.ndarray]:
```
Comment (heltluke) [inline strawberryfields/result.py:142]: Nice!

File: tests/api/test_result.py [hunk 13]
```diff
@@ -16,53 +16,122 @@
...
+
+
 class TestResult:
     """Tests for the ``Result`` class."""
 
+    def test_samples(self):
+        """Test that ``samples`` is correctly returned."""
+        result = Result(raw_results)
+        assert result.samples is not None
+        assert np.equal(result.samples, np.ones((2, 3, 4, 5))).all()
```
Comment (heltluke) [inline tests/api/test_result.py:40]: Why not
```suggestion
        assert np.array_equal(result.samples, np.ones((2, 3, 4, 5)))
```
?

File: tests/api/test_result.py [hunk 14]
```diff
@@ -16,53 +16,122 @@
...
 
 from strawberryfields.result import Result
 
 pytestmark = pytest.mark.api
 
 
+raw_results = {
+    "output": [np.ones((2, 3, 4, 5))],
+    "meta_array": np.array([1, 0, 1, 2]),
+    "meta_matrix": np.array([[1, 2], [3, 4]]),
```
Comment (heltluke) [inline tests/api/test_result.py:29]: Might make sense to assign the values (`[np.ones((2, 3, 4, 5))]`, `np.array([1, 0, 1, 2])`, `np.array([[1, 2], [3, 4]])`) to variables above, to reduce code repetition when making assertions below.

**Review by heltluke — APPROVED** [review]

File: strawberryfields/result.py (hunk 9 above)
Comment (heltluke) [reply strawberryfields/result.py:95]: Only able to be `list` after Python 3.9, right? So dependent on the Python versions supported?

File: strawberryfields/engine.py (hunk 8 above)
Comment (thisac) [reply strawberryfields/engine.py:226]: Oh, maybe. I basically added this to get rid of the type hint issue with calling `run_engine` from `BaseEngine._run`. Without it, my type-hinting engine assumes that `None` is being returned here, and thus complains. Perhaps I should just ignore it though. 🤔

File: strawberryfields/result.py (hunk 9 above)
Comment (thisac) [reply strawberryfields/result.py:95]: > Only able to be list after Python 3.9, right? So dependent on the Python versions supported?

I didn't know that Python 3.9 supported nested types in `list`. That's basically why I used `List` here instead, which is probably still a good idea if we'd like to have subtypes (although this technically only affects type hints, and not types in docstrings, I think).

> is the convention in PL and SF to use list or List in the Sphinx docstrings?

I don't think there is a convention here. `list` is definitely more common, but there are quite a few `List` occurrences as well across SF and PL.

File: strawberryfields/engine.py (hunk 10 above)
Comment (thisac) [reply strawberryfields/engine.py:399]: This only affects simulators, which return the results in the wrong order and thus need to be reshaped and transposed. So, if I recall correctly, this is needed for the results to have the same shape as the hardware.

File: tests/api/test_result.py (hunk 13 above)
Comment (thisac) [reply tests/api/test_result.py:40]: Good point, didn't think of it. 👍

File: strawberryfields/engine.py (hunk 10 above)
Comment (heltluke) [reply strawberryfields/engine.py:399]: I understand. I just find it a little odd that it's been left to SF to get results in the right shape, rather than demanding simulators send results to SF in the desired shape.

File: strawberryfields/result.py (hunk 1 above)
Comment (thisac) [reply strawberryfields/result.py:97]: Perhaps we should just remove it then? I see no real reason to have it, to be honest, since there will probably not be anything in `Result.raw` that isn't already provided via `Result.samples` or `Result.metadata`. 🤔

File: strawberryfields/engine.py (hunk 10 above)
Comment (thisac) [reply strawberryfields/engine.py:399]: Oh, I see. This is mostly due to the way that TDM programs are simulated. They need to be unrolled before being run, so what is _actually_ being simulated isn't specifically a TDM circuit. The backend itself doesn't know if it's working with a TDM program or not, and thus returns the measurements in the wrong order/on the wrong modes.

I think the `_run_program` method is the deepest we can put this transformation.

File: strawberryfields/result.py (hunk 9 above)
Comment (josh146) [reply strawberryfields/result.py:95]: > I don't think there is a convention here. list is definitely more common, but there are quite a few List occurrences as well across SF and PL.

Actually, this should be `list`, any instances of `List` were accidental or missed during code review! 

This is primarily because our Sphinx conventions predate Python type hinting

File: strawberryfields/result.py (hunk 11 above)
Comment (josh146) [reply strawberryfields/result.py:106]: Here our convention is `ndarray`, or in some cases, `array_like`, to match the numpy documentation

Comment (josh146) [reply strawberryfields/result.py:106]: e.g., https://numpy.org/doc/stable/reference/generated/numpy.dot.html#numpy.dot

File: strawberryfields/engine.py (hunk 10 above)
Comment (heltluke) [reply strawberryfields/engine.py:399]: Understood.

Very cool that the backend/simulator is agnostic to the program type.

File: strawberryfields/engine.py (hunk 8 above)
Comment (Mandrenkov) [reply strawberryfields/engine.py:226]: Oh, I see! In that case, I'm okay with either approach (although I guess it's better safe than sorry?).
